drivers/crypto/crypto_it8xxx2_sha_v2.c

```diff
@@ -0,0 +1,280 @@
+	gctrl_regs->GCTRL_PMER3;
+	irq_unlock(key);
+
+	chip_ctx.w_input_index = 0;
+}
+
+static int it8xxx2_hash_handler(struct hash_ctx *ctx, struct hash_pkt *pkt,
+				bool finish)
+{
+	struct gctrl_it8xxx2_regs *const gctrl_regs = GCTRL_IT8XXX2_REGS_BASE;
+	uint32_t rem_len = pkt->in_len;
+	uint32_t in_buf_idx = 0;
+	uint32_t i, key;
+	uint8_t sha_ctrl;
+
+	while (rem_len) {
+		/* Data length >= 1KB */
+		if (rem_len >= SHA_SHA256_SRAM_BUF) {
+			rem_len = rem_len - SHA_SHA256_SRAM_BUF;
+
+			for (i = 0; i < SHA_SHA256_SRAM_BUF; i++) {
+				chip_ctx.w_input[chip_ctx.w_input_index++] =
+								pkt->in_buf[in_buf_idx++];
+			}
+
+			/* HW automatically load 1KB data from DLM */
+			sys_write8(IT8XXX2_SHAEXEC_1KByte,
+					IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHAECR);
+			it8xxx2_sha256_module_calculation();
+		} else {
+			/* 0 <= Data length < 1KB */
+			while (rem_len) {
+				rem_len--;
+				chip_ctx.w_input[chip_ctx.w_input_index++] =
+								pkt->in_buf[in_buf_idx++];
+
+				if (chip_ctx.w_input_index >= SHA_SHA256_BLOCK_LEN) {
+					/* HW automatically load 64Bytes data from DLM */
+					sys_write8(IT8XXX2_SHAEXEC_64Byte,
+							IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHAECR);
+					it8xxx2_sha256_module_calculation();
+				}
+			}
+		}
+	}
+
+	chip_ctx.total_len += pkt->in_len;
+
+	if (finish) {
+		uint32_t *ob_ptr = (uint32_t *)pkt->out_buf;
+
+		/* Pre-processing (Padding) */
+		memset(&chip_ctx.w_input[chip_ctx.w_input_index],
+			0, SHA_SHA256_BLOCK_LEN - chip_ctx.w_input_index);
+		chip_ctx.w_input[chip_ctx.w_input_index] = 0x80;
+
+		if (chip_ctx.w_input_index >= 56) {
+			/* HW automatically load 64Bytes data from DLM */
+			sys_write8(IT8XXX2_SHAEXEC_64Byte,
+					IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHAECR);
+			it8xxx2_sha256_module_calculation();
+			memset(&chip_ctx.w_input[chip_ctx.w_input_index],
+				0, SHA_SHA256_BLOCK_LEN - chip_ctx.w_input_index);
+		}
+
+		/*
+		 * Since input data (big-endian) are copied 1byte by 1byte to
+		 * it8xxx2 memory (little-endian), so the bit length needs to
+		 * be transformed into big-endian format and then write to memory.
+		 */
+		chip_ctx.w_sha[15] = sys_cpu_to_be32(chip_ctx.total_len * 8);
+
+		/* HW automatically load 64Bytes data from DLM */
+		sys_write8(IT8XXX2_SHAEXEC_64Byte, IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHAECR);
+		it8xxx2_sha256_module_calculation();
+
+		/* HW write back the hash result to DLM */
+		/* Set DLM address for input data */
+		sys_write8(((uint32_t)&chip_ctx.h) & 0xc0,
+				IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHADBA0R);
+		sys_write8(((uint32_t)&chip_ctx.h) >> 8,
+				IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHADBA1R);
+
+		key = irq_lock();
+		/* Crypto use SRAM */
+		gctrl_regs->GCTRL_PMER3 |= IT8XXX2_GCTRL_SRAM_CRYPTO_USED;
+		sha_ctrl = sys_read8(IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHACR);
+		sys_write8(sha_ctrl | IT8XXX2_SHAWB, IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHACR);
+
+		while (sys_read8(IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHASR) & IT8XXX2_SHABUSY)
+			;
```

Comment:
unbounded loop

Reply:
I change the while(); loop to finite for loop with a timeout message.